File: 0033_searchInRotatedSortedArray/main.py
from typing import List

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l, r = 0, len(nums) - 1
        while l <= r:
            m = l + (r - l) // 2
            if nums[m] == target:
                return m
            
            # left sorted portion
            if nums[l] <= nums[m]:
                if target > nums[m] or target < nums[l]:
                    l = m + 1
                else:
                    r = m - 1
            # right sorted portion
            else:
                if target < nums[m] or target > nums[r]:
                    r = m - 1
                else:
                    l = m + 1
        return -1
    

class Solution1:
    def search(self, nums: List[int], target: int) -> int:
        l, r = 0, len(nums) - 1
        while l <= r:
            m = l + (r - l) // 2
            if nums[m] == target:
                return m
            
            # left sorted portion
            if nums[l] <= nums[m]:
                if target < nums[m] and target >= nums[l]:
                    r = m - 1
                else:
                    l = m + 1
            # right sorted portion
            else:
                if target > nums[m] and target <= nums[r]:
                    l = m + 1
                else:
                    r = m - 1
        return -1
    
if __name__ == "__main__":
    print(Solution1().search([4,5,6,7,0,1,2], 0))

File: 0033_searchInRotatedSortedArray/main.cpp
#include <vector>

class Solution {
public:
    int search(std::vector<int>& nums, int target) {
        int l, r, m;
        l = 0;
        r = nums.size() - 1;

        while (l <= r) {
            m = l + (r - l) / 2;
            if (nums[m] == target)
                return m;

            // left portion sorted
            if (nums[m] >= nums[l]) {
                if (target >= nums[l] && target < nums[m])
                    r = m - 1;
                else
                    l = m + 1;
            } else {  // right portion sorted
                if (target > nums[m] && target <= nums[r])
                    l = m + 1;
                else
                    r = m - 1;
            }
        }
        return -1;
    }
};